File: cpp/july/day_27_Construct Binary Tree from Inorder and Postorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* build(vector<int>& in, vector<int>& post, int l, int r, int& post_idx){
        if(l > r){
            return NULL;
        }
        int idx = 0;
        while(in[idx] != post[post_idx]){
            idx++;
        }
        TreeNode* root = new TreeNode(post[post_idx--]);
        if(l == r){
            return root;
        }
        root->right = build(in, post, idx+1, r, post_idx);
        root->left = build(in, post, l, idx-1, post_idx);
        
        return root;
    }
    
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if(postorder.size() == 0){
            return NULL;
        }
        int post_idx = postorder.size() - 1;
        return build(inorder, postorder, 0, inorder.size() - 1, post_idx);
    }
};

File: python3/july/day_27_Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def build(self, inorder, postorder, l, r):
        if(l > r):
            return None

        idx = 0
        while(inorder[idx] != postorder[self.post_idx]):
            idx += 1
        
        root = TreeNode(postorder[self.post_idx])
        self.post_idx -= 1
        if(l == r):
            return root

        root.right = self.build(inorder, postorder, idx+1, r)
        root.left = self.build(inorder, postorder, l, idx-1)
        
        return root

    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if len(postorder) == 0:
            return None
        
        self.post_idx = len(postorder) - 1
        return self.build(inorder, postorder, 0, len(inorder) - 1)
